### PR TITLE
Added stateful.layers.Embedding and test function

### LINKED ISSUE #16496: embedding

### PR DESCRIPTION
**Pull request by baberabb**
closes #16496 

Tried to include most of the parameters from [torch](https://pytorch.org/docs/stable/generated/torch.nn.Embedding.html) and [tensorflow](https://www.tensorflow.org/api_docs/python/tf/keras/layers/Embedding), however they are not implemented in the backend functions (all experimental).

### CODE REVIEW

**Comment by ivy-leaves** [comment]
If you are working on an open task, please **edit** the PR description to link to the issue you've created.

For more information, please check [ToDo List Issues Guide](https://unify.ai/docs/ivy/overview/contributing/the_basics.html#todo-list-issues).

Thank you :hugs:

**Comment by ivy-leaves** [comment]
If you are working on an open task, please **edit** the PR description to link to the issue you've created.

For more information, please check [ToDo List Issues Guide](https://unify.ai/docs/ivy/overview/contributing/the_basics.html#todo-list-issues).

Thank you :hugs:

File: ivy_tests/test_ivy/helpers/hypothesis_helpers/general_helpers.py [hunk 1]
```diff
@@ -527,9 +527,7 @@ def embedding_helper(draw):
     dtype_weight, weight = draw(
         array_helpers.dtype_and_values(
             available_dtypes=[
-                x
-                for x in draw(dtype_helpers.get_dtypes("numeric"))
-                if "float" in x or "complex" in x
+                x for x in draw(dtype_helpers.get_dtypes("numeric")) if "float" in x
```
Comment (CatB1t) [inline ivy_tests/test_ivy/helpers/hypothesis_helpers/general_helpers.py:530]: Hello @baberabb, why we did we remove `complex` from the here? also if we only want `float` dtypes, we can use `draw(dtype_helpers.get_dtypes("float")` right away.

File: ivy_tests/test_ivy/helpers/hypothesis_helpers/general_helpers.py [hunk 2]
```diff
@@ -549,4 +547,11 @@ def embedding_helper(draw):
...
     padding_idx = draw(st.integers(min_value=0, max_value=num_embeddings - 1))
-    return dtype_indices + dtype_weight, indices[0], weight[0], padding_idx
+    return (
+        dtype_indices + dtype_weight,
+        indices[0],
+        weight[0],
+        padding_idx,
+        num_embeddings,
+        embedding_dim,
+    )
```
Comment (CatB1t) [inline ivy_tests/test_ivy/helpers/hypothesis_helpers/general_helpers.py:557]: Let's make sure the new returns doesn't break old usage of the embedding helper.

**Review by CatB1t — CHANGES_REQUESTED** [review]

File: ivy_tests/test_ivy/helpers/hypothesis_helpers/general_helpers.py (hunk 1 above)
Comment (baberabb) [reply ivy_tests/test_ivy/helpers/hypothesis_helpers/general_helpers.py:530]: Hi @CatB1t You're right, using `draw(dtype_helpers.get_dtypes("float"))` directly would have been better. I removed `complex` because embeddings typically only use real weights. Will change it back.

**Comment by baberabb** [comment]
Hi @CatB1t @DragosStoican. So i added a `st.composite` function to the testing implementation rather than amending the `embedding_helper`. Hope this is better!

File: ivy_tests/test_ivy/test_stateful/test_layers.py [hunk 3]
```diff
@@ -1378,3 +1378,81 @@ def test_adaptive_avg_pool1d_layer(
...
+            available_dtypes=helpers.get_dtypes("float"),
+            min_num_dims=2,
+            max_num_dims=2,
+            min_dim_size=1,
+        )
+    )
+    num_embeddings, embedding_dim = weight[0].shape
+    dtype_indices, indices = draw(
+        helpers.dtype_and_values(
+            available_dtypes=["int32", "int64"],
```
Comment (CatB1t) [inline ivy_tests/test_ivy/test_stateful/test_layers.py:1396]: We should be ideally testing for all valid dtypes, and not only a subset.

File: ivy_tests/test_ivy/test_stateful/test_layers.py [hunk 4]
```diff
@@ -1378,3 +1378,81 @@ def test_adaptive_avg_pool1d_layer(
         test_gradients=test_gradients,
         on_device=on_device,
     )
+
+
+@st.composite
+def _embedding_helper(draw):
+    dtype_weight, weight = draw(
+        helpers.dtype_and_values(
+            available_dtypes=helpers.get_dtypes("float"),
```
Comment (CatB1t) [inline ivy_tests/test_ivy/test_stateful/test_layers.py:1387]: We should be ideally testing for all valid dtypes, and not only a subset.

File: ivy_tests/test_ivy/test_stateful/test_layers.py [hunk 5]
```diff
@@ -1378,3 +1378,81 @@ def test_adaptive_avg_pool1d_layer(
...
+    )
+    num_embeddings, embedding_dim = weight[0].shape
+    dtype_indices, indices = draw(
+        helpers.dtype_and_values(
+            available_dtypes=["int32", "int64"],
+            min_num_dims=2,
+            min_dim_size=1,
+            min_value=0,
+            max_value=num_embeddings - 1,
+        ).filter(lambda x: x[1][0].shape[-1] == embedding_dim)
```
Comment (CatB1t) [inline ivy_tests/test_ivy/test_stateful/test_layers.py:1401]: Let's add a comment for clarity of the purpose of the filtering, for future code readers 😄

File: ivy_tests/test_ivy/test_stateful/test_layers.py [hunk 6]
```diff
@@ -1378,3 +1378,81 @@ def test_adaptive_avg_pool1d_layer(
...
+            max_num_dims=2,
+            min_dim_size=1,
+        )
+    )
+    num_embeddings, embedding_dim = weight[0].shape
+    dtype_indices, indices = draw(
+        helpers.dtype_and_values(
+            available_dtypes=["int32", "int64"],
+            min_num_dims=2,
+            min_dim_size=1,
```
Comment (CatB1t) [inline ivy_tests/test_ivy/test_stateful/test_layers.py:1398]: Are those restrictions required? ideally we should allow for the search space to expand.

**Review by CatB1t — CHANGES_REQUESTED** [review]

**Comment by ivy-seed** [comment]
This PR has been labelled as stale because it has been inactive for more than 7 days. If you would like to continue working on this PR, then please add another comment or this PR will be closed in 7 days.

**Comment by ivy-seed** [comment]
This PR has been closed because it has been marked as stale for more than 7 days with no activity.
